### PR TITLE
adding a general detection object, making BaseMap and BaseObjectDetec…

### PR DESCRIPTION
**Pull request by Leibovitch**
…tion extend it. adding string to tile id

### CODE REVIEW

File: analona.py [hunk 1]
```diff
@@ -19,126 +23,185 @@ def validate(self):
...
+            }
+        }
+
+    @staticmethod
+    def compose_schema(schema):
+        schema.update(BaseDetection.unique_parameters)
+        return schema
+
+    def __init__(self, item, schema = {}):
+        print('\n base detector costructor called')
```
Comment (taihak94) [inline analona.py]: remove

File: analona.py [hunk 2]
```diff
@@ -19,126 +23,185 @@ def validate(self):
...
+            },
+            Optional('linkToSourceObjext'): {
+                "url": Or(Regex(url_regex), Regex(storage_regex), error="analyticsInfo: invalid url"),
+                "storage": Or("Azure", "AWS", "GoogleCloud", "Planet", error="analyticsInfo: unknown storage type")
+            }
+        }
+
+    @staticmethod
+    def compose_schema(schema):
+        schema.update(BaseDetection.unique_parameters)
```
Comment (taihak94) [inline analona.py]: if you use `BaseDetection` then it shouldn't really be static, but should be self, no?

File: analona.py [hunk 3]
```diff
@@ -19,126 +23,185 @@ def validate(self):
...
+        return schema
+
+    def __init__(self, item, schema = {}):
+        print('\n base detector costructor called')
+        Validator.__init__(self, item, BaseDetection.compose_schema(schema))
```
Comment (taihak94) [inline analona.py]: use `self.compose_schema`

File: analona.py [hunk 4]
```diff
@@ -19,126 +23,185 @@ def validate(self):
...
+
+
+
+class BaseObject(BaseDetection):
+    """
+    General valiator for single item (Ship/Plane)
```
Comment (taihak94) [inline analona.py]: validator

File: analona.py [hunk 5]
```diff
@@ -19,126 +23,185 @@ def validate(self):
...
             'originalImageId': And(str, len),
-            'observed_start': datetime, 
-            'observed_end': datetime,
             'area': Or(float, int),
             'length': Or(float, int),
             'width': Or(float, int),
             Optional('score'): And(Use(float), lambda s: 0 <= s <= 1,error="score: should be between 0-1"),
             Optional('direction'): And(Use(float), lambda s: 0 <= s <= 360, error="direction: should be between 0-360")
-        })
+
```
Comment (taihak94) [inline analona.py]: remove extra space

File: analona.py (hunk 2 above)
Comment (Leibovitch) [reply analona.py]: no, all objects are of the same structure except the Validator. Base detection is NOT the top of the hierarchy, and thus it's method is static as well

File: analona.py (hunk 3 above)
Comment (Leibovitch) [reply analona.py]: see previous comment

File: analona.py [hunk 6]
```diff
@@ -19,126 +20,179 @@ def validate(self):
...
+        }
+    }
+
+    @staticmethod
+    def compose_schema(schema):
+        schema.update(BaseDetection.unique_parameters)
+        return schema
+
+    def __init__(self, item, schema = {}):
+        super().__init__(item, BaseDetection.compose_schema(schema))
```
Comment (taihak94) [inline analona.py]: 👏

File: analona.py [hunk 7]
```diff
@@ -19,126 +20,179 @@ def validate(self):
...
             'area': Or(float, int),
             'length': Or(float, int),
             'width': Or(float, int),
             Optional('score'): And(Use(float), lambda s: 0 <= s <= 1,error="score: should be between 0-1"),
             Optional('direction'): And(Use(float), lambda s: 0 <= s <= 360, error="direction: should be between 0-360")
-        })
+            
+        }
+
+
```
Comment (taihak94) [inline analona.py]: remove double space

File: analona.py [hunk 8]
```diff
@@ -19,126 +20,179 @@ def validate(self):
...
-        BaseMap.__init__(self, building)
+    unique_parameters = {}
+
+    @staticmethod
+    def compose_schema(schema):
+        schema.update(BaseMap.unique_parameters)
+        return schema
+
+    def __init__(self, item, schema = {}):
+        BaseMap.__init__(self, item, Building.compose_schema(schema))
```
Comment (taihak94) [inline analona.py]: super

File: test/buildingTest.py [hunk 9]
```diff
@@ -1,5 +1,8 @@
 from datetime import datetime
 import os,sys,inspect
+import pprint
+
+pp = pprint.PrettyPrinter(indent=4)
```
Comment (taihak94) [inline test/buildingTest.py]: remove pprint

File: test/shipsTest.py [hunk 10]
```diff
@@ -1,5 +1,8 @@
 from datetime import datetime
 import os,sys,inspect
+import pprint
+
+pp = pprint.PrettyPrinter(indent=4)
```
Comment (taihak94) [inline test/shipsTest.py]: remove pprint

File: test/shipsTest.py [hunk 11]
```diff
@@ -48,6 +51,7 @@
     }
 }
 
+
```
Comment (taihak94) [inline test/shipsTest.py]: remove extra line

File: analona.py (hunk 4 above)
Comment (taihak94) [reply analona.py]: fix it please

File: analona.py [hunk 12]
```diff
@@ -19,126 +20,180 @@ def validate(self):
...
     """
     Ship validator
     """
-    def __init__(self, ship):
-        BaseObject.__init__(self, ship)
+
+    unique_parameters = {
+        Optional('AISIdentifiers'): {
+            Optional('IMO'): Or(str, int),
+            Optional('IMO'): Or(str, int),
```
Comment (taihak94) [inline analona.py]: you have two IMO

File: test/allTest.py [hunk 13]
```diff
@@ -0,0 +1,20 @@
+import shipsTest
+import buildingTest
+import planesTest
+import roadsTest
+import vegetationTest
+
+print("TESTING BASE OBJECTS\n")
+print("############\n")
+print("testing ships:")
```
Comment (taihak94) [inline test/allTest.py]: remove the printsssssssssssssssss

File: test/shipsTest.py (hunk 10 above)
Comment (taihak94) [reply test/shipsTest.py]: please 🙏

File: analona.py [hunk 14]
```diff
@@ -19,126 +20,179 @@ def validate(self):
...
+            'update_time': datetime,
+            Optional('tile_id'): {
+                "row": Or(str, Use(int)),
+                "col": Or(str, Use(int)),
+                "full_id": str
+            },
+            Optional('linkToSourceObject'): {
+                "url": Or(Regex(url_regex), Regex(storage_regex), error="analyticsInfo: invalid url"),
+                "storage": Or("Azure", "AWS", "GoogleCloud", "Planet", error="analyticsInfo: unknown storage type")
+
```
Comment (taihak94) [inline analona.py]: remove extra newline

**Pull request merged**